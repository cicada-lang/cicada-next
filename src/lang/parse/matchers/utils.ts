import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
