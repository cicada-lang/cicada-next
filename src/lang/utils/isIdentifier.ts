const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
