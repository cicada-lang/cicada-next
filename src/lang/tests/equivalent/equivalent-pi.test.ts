import { expect, test } from "vitest"
import { expectCodeToFail, runCode } from "../utils"

test("equivalent Pi", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent Type {
    forall (T: Type, x: T) T
  = forall (T: Type, y: T) T
  = forall (T: Type, y: T) id(Type, T)
  = forall (T: Type, y: id(Type, T)) id(Type, T)
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Pi -- more nested", async () => {
  const output = await runCode(`

function id(T: Type, x: T): T {
  return x
}

compute equivalent Type {
    forall (A: Type, x: A, B: Type, y: B) A
  = forall (B: Type, y: B, A: Type, x: A) B
  = forall (B: Type, y: id(Type, B), A: Type, x: id(Type, A)) id(Type, B)
}

`)

  expect(output).toMatchSnapshot()
})

test("equivalent Pi -- fail", async () => {
  await expectCodeToFail(`

compute equivalent Type {
    forall (A: Type, x: A, B: Type, y: B) A
  = forall (A: Type, x: A, B: Type, y: B) B
}

`)
})
