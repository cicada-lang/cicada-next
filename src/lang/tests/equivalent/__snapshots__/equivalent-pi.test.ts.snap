// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`equivalent Pi -- more nested 1`] = `
"refl(implicit Type, implicit (A: Type, A, B: Type, B) -> A): Equal(
  Type,
  (A: Type, A, B: Type, B) -> A,
  (B: Type, B, A: Type, A) -> B,
)"
`;

exports[`equivalent Pi 1`] = `"refl(implicit Type, implicit (T31: Type, T31) -> T31): Equal(Type, (T5: Type, T5) -> T5, (T5: Type, T5) -> T5)"`;
